relayer/send_relayed_transaction.go

```diff
@@ -0,0 +1,84 @@
+package relayer
+
+import (
+	"context"
+
+	"github.com/athanorlabs/go-relayer/impls/gsnforwarder"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+	"github.com/athanorlabs/atomic-swap/ethereum/block"
+	"github.com/athanorlabs/atomic-swap/ethereum/extethclient"
+	"github.com/athanorlabs/atomic-swap/net/message"
+)
+
+// ValidateAndSendTransaction sends the relayed transaction to the network if it validates successfully.
+func ValidateAndSendTransaction(
+	ctx context.Context,
+	req *message.RelayClaimRequest,
+	ec extethclient.EthClient,
+	ourSFContractAddr ethcommon.Address,
+) (*message.RelayClaimResponse, error) {
+
+	err := validateClaimRequest(ctx, req, ec.Raw(), ourSFContractAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	reqSwapFactory, err := contracts.NewSwapFactory(req.SFContractAddress, ec.Raw())
+	if err != nil {
+		return nil, err
+	}
+
+	reqForwarderAddr, err := reqSwapFactory.TrustedForwarder(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		return nil, err
+	}
+
+	reqForwarder, domainSeparator, err := getForwarderAndDomainSeparator(ctx, ec.Raw(), reqForwarderAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce, err := reqForwarder.GetNonce(&bind.CallOpts{Context: ctx}, req.Swap.Claimer)
+	if err != nil {
+		return nil, err
+	}
+
+	// The size of request.Secret was vetted when it was deserialized
+	secret := (*[32]byte)(req.Secret)
+
+	// Lock the wallet's nonce until we get a receipt
+	ec.Lock()
+	defer ec.Unlock()
```

Comment:
move this to above the `ec.TxOpts(ctx)` line?

Reply:
I moved the `ec.Lock()` call to after `createForwarderRequest(...)` which didn't need the lock's protection. That may have been what you were asking for. I left `ec.TxOpts(ctx)` inside the lock, because it's a fast/trivial operation and leaving it inside the lock unburdens the reader from needing to know that it leaves the `Nonce` field of the returned object as `nil`. (Otherwise, it would need to be inside the lock.)